Restore EAX effects through OpenAL's EFX.

Deledrius opened the pull request:
This re-implements the supported EAX effects using OpenAL's EFX.  Some features (such as occlusion effects) are currently unsupported by OpenAL and required conversion or local implementation.

patmauro commented:
If there's any interest in dusting this one off again, I would love to help in any way I can, depending on what the current blockers may be. This feature in particular is very near and dear to me - URU has never _quite_ felt like URU without it, and getting this functionality restored even _partially_ would be a tremendously huge W. So again, please let me know if I can be of any assistance here; my C++ is a little rusty, but I'm certainly no stranger to reviewing scripts and running tests.

Here's a thought - on the front-end side, in addition to the simple change of _"Enable EAX"_ to _"Enable E**F**X"_ in the options menu (there doesn't really seem to be room in the GUI for a description more detailed than that, unfortunately), we'll probably want to replace that "_EAX ADVANCED HD_" graphic at the bottom of the audio menu with a more generic OpenAL one as well:
![eax_to_openal](https://github.com/H-uru/Plasma/assets/6642931/451176a3-2b00-490c-88d2-af48324c1cfa)

This would require a modification to the **xoptionsguicomponents** texture contained inside **GUI_District_OptionsMenuGUI.prp** - we don't have a source for this one (the version of this we have in the source files is from CC and is quite different), so as usual, the trick here would be figuring out how to modify this texture without introducing any new compression artifacts to the rest of the texture sheet. _(**Granted - this wouldn't be covered by this PR and would require a separate PR to moul-assets**, but it seemed relevant to mention here.)_

Deledrius commented:
No dusting off necessary, I've been actively working on this.

patmauro commented:
> No dusting off necessary, I've been actively working on this.
> 
> Updating the in-game text and graphics will definitely be something we'll need. I'll let you know when I get close to that stage!

Sounds good. 🙂 In any case, the offer stands; happy to make myself available if you need an extra set of hands. 

I may go ahead and look into a PR for that graphic in _moul-assets_ unless there's any objection... seems like it might not be a bad idea to start ditching the user-facing EAX references anyway, regardless of the status of the EFX project. 🤔

Deledrius commented:
Sure, go ahead. We can replace that texture with a lossless one if it isn't already, so additional edits won't introduce any generational loss.  I can't check on this at the moment, but feel free to start on a plan for it!

dpogue commented on Sources/Plasma/PubUtilLib/plAudio/plAudioSystem.cpp:
```diff
@@ -41,12 +41,7 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 *==LICENSE==*/
 #include "HeadSpin.h"
 #include <al.h>
-#ifdef USE_EFX
-#   include <efx.h>
-#endif
```
To slightly complicate things, there are some OpenAL implementations (Apple) that don't ship the EFX header, which is why this was conditionally detected.

All of our builds use vcpkg which uses openal-soft, which includes EFX, but it would be nice to still gracefully fall back to no effects if EFX can't be found

Deledrius replied:
Ah, that's unfortunate.  I did this assuming we'd be using the same library across the different platforms, where OpenAL-soft provides this for us without variability.  

It does complicate things, but I can see about re-introducing proper conditional support.

dpogue replied:
If it turns out to be a mess, we're probably okay to mandate OpenAL-Soft.

It's just nice to fallback to system libraries on weird old systems where you don't feel like trying to compile extra dependencies 😛

Deledrius replied:
Yeah, it's a good idea.  I just assumed I could simplify that code a bit and took the opportunity to clean it out. ;)

dpogue reviewed (commented):
Does it seem feasible to wrap this in USE_EFX conditionally, or is that going to turn into a mess of ifdefs everywhere? The `AL_AUXILIARY_SEND_FILTER` stuff seems like the worst in terms of wrapping, but maybe we can turn that into a conditional macro?

patmauro commented:
Loading in this latest draft and testing it out put an enormous smile on my face. You've done something important here, @Deledrius, and should feel very proud.

One note - since the `Audio.UseEAX` syntax has been changed to `Audio.UseEFX`, the app will throw an error if an existing audio.ini config file exists in local AppData with the old syntax - even if you "press OK to continue parsing files," it seemingly _won't_, and will launch in 800x600 with no sound - I had to purge the existing `(username)/AppData/Local/Uru Live/` dir and let it regenerate to get it to cooperate. May not be a bad idea to gracefully handle the old ini syntax to account for this (even if it just specifically calls this out as something to *ignore*).

dpogue commented:
> One note - since the `Audio.UseEAX` syntax has been changed to `Audio.UseEFX`, the app will throw an error if an existing audio.ini config file exists in local AppData with the old syntax

I was thinking about this and almost left a comment when I reviewed the code last night, but thought "EAX hasn't worked for years, how likely is it that anyone even has MOULa config files with that option enabled?"

I guess it is a possibility, and maybe supporting both commands (or at least support parsing the `UseEAX` one with no implementation) might be best.

Deledrius commented:
> "EAX hasn't worked for years, how likely is it that anyone even has MOULa config files with that option enabled?"

That was my thinking, too, when I changed it; but I'm fairly certain I changed code that's been writing out the old settings (as false) all this time, so the answer is probably "everyone".

Sounds like the correct solution here is to figure out why we're not ignoring it gracefully and fix that.  It should simply be able to ignore an unknown value and continue without nuking everything.

Hazado commented:
Maybe blowing up here?

https://github.com/H-uru/Plasma/blob/master/Scripts/Python/xIniAudio.py#L290

Hoikas commented:
Given that we have lots of config files running around out there that use the command `Audio.UseEAX`, I think it's critical that the command remain named thusly. I realize that this is a someone inaccurate name, but sometimes we have to make concessions for ease of use. Further, we can't really control what client the user is running at all, so it's important that we don't start dumping `Audio.UseEFX` into files as well. If the user runs an old client, and the file contains that command, their old client will error. This will even happen on even slightly out of date H'uru clients. We don't want that.

We don't want to remove this error either. That error is coming from the `pfConsole` command parser. We want the console to tell us when it receives an invalid input. We don't want it to simply drop unknown commands on the floor. This is an indication that something is wrong. Like someone has edited the FNI file and misspelled something.

I think I would actually go a step farther than this and say since the classes are all still named `plEAXWhatever`, and we're leaving the console command as `Audio.UseEAX`, that we shouldn't change any of the engine's function or variable names at all to `EFX` from `EAX`. While this change does clarify the underlying backend for environmental audio effects, in practice it muddies the water significantly for anyone who wants to maintain something like a Python script for both TPotS and H'uru. However, we should *definitely* update the user facing strings to refer to EFX.

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.cpp:
```diff
@@ -287,21 +212,18 @@ void    plEAXListener::ClearProcessCache()
-    static int oldTime = timeGetTime();     // Get starting time
-    int newTime;
+    EFXEAXREVERBPROPERTIES  finalProps;
+    static auto oldTime = std::chrono::steady_clock::now();     // Get starting time
```
Bump

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.cpp:
```diff
@@ -411,55 +329,41 @@ void    plEAXListener::ProcessMods(const std::set<plEAXListenerMod*>& modArray )
     if(bMorphing)
     {
-        newTime = timeGetTime();
+        auto newTime = std::chrono::steady_clock::now();
```
Bump

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.cpp:
```diff
@@ -411,55 +329,41 @@ void    plEAXListener::ProcessMods(const std::set<plEAXListenerMod*>& modArray )
         // Update, at most, ten times per second
-        if((newTime - oldTime) < 100) return;
-            
+        if((newTime - oldTime) < std::chrono::duration<int>(100))
```
Bump

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.cpp:
```diff
@@ -666,55 +556,66 @@ bool    plEAXSource::IsValid() const
-#if EAX_SDK_AVAILABLE
     uint32_t dirtyParams;
-    if(source == 0 || !fInit) 
+    if(source == 0 || !fInit)
```
Bump

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.cpp:
```diff
@@ -666,55 +556,66 @@ bool    plEAXSource::IsValid() const
         return;
 
-    if( force )
+    if(force)
```
Bump

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.cpp:
```diff
@@ -666,55 +556,66 @@ bool    plEAXSource::IsValid() const
     }
 
-    if( dirtyParams & plEAXSourceSettings::kOutsideVolHF )
+    if(dirtyParams & plEAXSourceSettings::kOutsideVolHF)
```
Bump

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.cpp:
```diff
@@ -666,55 +556,66 @@ bool    plEAXSource::IsValid() const
     }
 
-    if( dirtyParams & plEAXSourceSettings::kOcclusion )
+    if(dirtyParams & plEAXSourceSettings::kOcclusion)
```
Bump

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.h:
```diff
@@ -49,58 +50,46 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 #ifndef _plEAXEffects_h
 #define _plEAXEffects_h
 
+#include <set>
+#include <efx.h>
+#include <efx-presets.h>
```
It would be nice to avoid pulling these library headers into a one of our headers. Are there any forward declarations we can use to avoid this?

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.h:
```diff
@@ -134,8 +122,8 @@ class plEAXSource;
-        plEAXSourceSettings();
-        virtual ~plEAXSourceSettings();
+        plEAXSourceSettings() : fDirtyParams(kAll) { Enable(false); }
+        virtual ~plEAXSourceSettings() { }
```
Bump

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXEffects.h:
```diff
@@ -201,18 +189,16 @@ class plEAXSource
 
-    plEAXSource();
-    virtual ~plEAXSource();
+    plEAXSource() : fInit(false) {}
```
Bump

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXListenerMod.cpp:
```diff
@@ -208,45 +205,135 @@ void plEAXListenerMod::Write( hsStream* s, hsResMgr* mgr )
+
+/* Helper functions for converting to EFX from deprecated EAX values */
+
+#include <math.h>
```
Let's move this include to the top of the file.

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXListenerMod.cpp:
```diff
@@ -208,45 +205,135 @@ void plEAXListenerMod::Write( hsStream* s, hsResMgr* mgr )
+    return 2000.0f * log10f(gain);
+}
+
+LPEFXEAXREVERBPROPERTIES ConvertEAXToEFX(const LPEAXLISTENERPROPERTIES eax)
```
```suggestion
static LPEFXEAXREVERBPROPERTIES ConvertEAXToEFX(const LPEAXLISTENERPROPERTIES eax)
```

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXListenerMod.cpp:
```diff
@@ -208,45 +205,135 @@ void plEAXListenerMod::Write( hsStream* s, hsResMgr* mgr )
+    return efx;
+}
+
+LPEAXLISTENERPROPERTIES ConvertEFXToEAX(const LPEFXEAXREVERBPROPERTIES efx)
```
```suggestion
static LPEAXLISTENERPROPERTIES ConvertEFXToEAX(const LPEFXEAXREVERBPROPERTIES efx)
```

Hoikas commented on Sources/Plasma/PubUtilLib/plAudio/plEAXListenerMod.h:
```diff
@@ -51,15 +52,13 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 
 #include "pnModifier/plSingleModifier.h"
 
+#include <efx.h>
+#include <efx-presets.h>
```
It would be nice to avoid including this library header in our headers. Are there any forward declarations we can use?

Hoikas reviewed (commented):
Some additional review items on top of the concerns from the PR comment about naming. Also, it would be nice if we could cull the no-op whitespace changes to make reviewing easier.

patmauro commented:
> Given that we have lots of config files running around out there that use the command `Audio.UseEAX`, I think it's critical that the command remain named thusly. I realize that this is a someone inaccurate name, but sometimes we have to make concessions for ease of use.

Just FYI, I noticed that the "old" ini I cleared out appeared to already have entries for _both_ `Audio.UseEAX` _and_ `Audio.UseEFX` - so it seems like this had _already_ been happily working in parallel prior to this PR, and something about the specific changes here may have simply rendered the "EAX" syntax invalid - not to mention, it seems to suggest that the config file seems to have made it through the switchover to the "EFX" syntax with no issues, so long as the legacy term was still considered valid.

So it _might_ be sufficient to revert whatever did that, but - to your point, maybe consistently using the legacy term (at least internally) would be best. In any case, just wanted to provide that additional context - I trust you guys will determine what makes the most sense here. 😄

patmauro commented:
> Just FYI, I noticed that the "old" ini I cleared out appeared to already have entries for _both_ `Audio.UseEAX` _and_ `Audio.UseEFX` - so it seems like this had _already_ been happily working in parallel prior to this PR, and something about the specific changes here may have simply rendered the "EAX" syntax invalid

Actually, disregard this - I just fired up a fresh Huru client and it seemed to have as much trouble with the `Audio.UseEFX` field as the build that includes this PR had with the `Audio.UseEAX` field. So yeah, I think Hoikas is right and this field will probably need to stay as "EAX" to ensure compatibility.

Deledrius commented:
Thankfully, the conflicts introduced by the Python updates are just a few cosmetic changes in the rename from EAX to EFX and are easy to resolve when I'm at my system.  

This is almost complete, AFAIK.  Some values need testing and checking, and the source properties need attention before it's as close to 100% as I can get it.
